brian2/tests/test_neurongroup.py

```diff
@@ -1206,6 +1206,95 @@ def test_get_states():
     assert set(all_states.keys()) == {'v', 'x', 'N', 't', 'dt', 'i',
                                       'subexpr', 'subexpr2'}
 
+@attr('codegen-independent')
+def test_set_states():
+    G = NeuronGroup(10, '''v : volt
+                           x : 1
+                           subexpr = x + v/volt : 1
+                           subexpr2 = x*volt + v : volt''')
+    G.v = 'i*volt'
+    G.x = '10*i'
+    assert_raises(ValueError, lambda: G.set_states({'v': np.arange(2,11)*volt}, units=True))
+    # we test if function prevents from setting read_only variables
+    assert_raises(TypeError, lambda: G.set_states({'N': 1}))
+    assert_raises(DimensionMismatchError, lambda: G.set_states({'x': np.arange(2,12)*volt}, units=True))
+    assert_raises(DimensionMismatchError, lambda: G.set_states({'v': np.arange(2,12)}, units=True))
+    G.set_states({'v':np.arange(2,12)}, units=False)
+    assert_equal(G.v, np.arange(2,12)*volt)
+    G.set_states({'v':np.arange(2,12)*volt}, units=True)
+    assert_equal(G.v, np.arange(2,12)*volt)
+    G.set_states({'x':np.arange(2,12)}, units=False)
+    assert_equal(G.x, np.arange(2,12))
+    G.set_states({'x':np.arange(2,12)}, units=True)
+    assert_equal(G.x, np.arange(2,12))
+
+@attr('codegen-independent')
+def test_get_states_pandas():
+    try:
+        import pandas as pd
+    except ImportError:
+        raise SkipTest('Cannot test export to Pandas data frame, Pandas is not installed.')
+    G = NeuronGroup(10, '''v : volt
+                           x : 1
+                           subexpr = x + v/volt : 1
+                           subexpr2 = x*volt + v : volt''')
+    G.v = 'i*volt'
+    G.x = '10*i'
+    states_units = G.get_states(['v', 'x', 'subexpr', 'subexpr2'], units=True, format='pandas')
+    states = G.get_states(['v', 'x', 'subexpr', 'subexpr2'], units=False, format='pandas')
+    assert len(states_units.columns) == len(states.columns) == 4
+    assert_equal(states_units['v'].as_matrix(), np.arange(10))
```

Comment:
A minor issue: I think the use of `.as_matrix()` is deprecated (it is confusing because it returns a numpy array, not a numpy matrix...), you can use `.values` instead.